hydra_provider/src/data_feed_service/serial/service.rs

```diff
@@ -0,0 +1,103 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use tokio::sync::Mutex;
+use messages::mavlink::connect;
+use serialport::{available_ports, SerialPortType};
+use tonic::{async_trait, Request, Response, Status};
+use log::info;
+
+use crate::data_feed_service::serial::iterator::SerialDataFeedIterator;
+use crate::database_service::DatabaseService;
+use crate::data_feed_service::proto::serial_data_feed_server::*;
+use crate::data_feed_service::proto::{Empty, ListAvailablePortsResponse, SerialDataFeedConfig, SerialDataFeedStatus};
+
+use messages::mavlink::uorocketry::MavMessage;
+
+#[derive(Clone)]
+pub struct SerialDataFeedService {
+	iterator: SerialDataFeedIterator,
+	database_service: Arc<Mutex<DatabaseService>>,
+}
+
+impl SerialDataFeedService {
+	pub fn new(
+		database_service: Arc<Mutex<DatabaseService>>,
+	) -> SerialDataFeedService {
+		SerialDataFeedService {
+			iterator: SerialDataFeedIterator {
+				is_running: Arc::new(AtomicBool::new(false)),
+				mavlink: Arc::new(Mutex::new(None)),
+				config: Arc::new(Mutex::new(None))
+			},
+			database_service,
+		}
+	}
+}
+
+#[async_trait]
+impl SerialDataFeed for SerialDataFeedService {
+	async fn start(&self, _request: Request<Empty>) -> Result<Response<Empty>, Status> {
+		self.iterator.is_running.store(true, Ordering::Relaxed);
+
+		let mut iterator = self.iterator.clone();
+		let database_service = self.database_service.clone();
+		tokio::spawn(async move {
+			info!("SerialDataFeedService has started.");
+			while let Some(message) = iterator.next().await {
+				// SHOULD DO: figure out how to handle database save errors better
+				let _ = database_service.lock().await.save(message).await;
```

Comment:
I'm not sure how I feel about the database being behind a lock. I suppose lock contention shouldn't be an issue. Only thing is since there is no buffer, any failure or degredation of the database (say it's being slow) could possibly mean some messages from the rocket are dropped.

What do you think of having a separate database task? `SerialDataFeedService` and `RandomDataFeedService` would send data to it via a bounded channel. The task would be dead simple and probably live in bootstrap, just reading from the channel and saving to database. Bonus, central location to handle those possible errors while saving.

Reply:
Yeah I'll add that in the issues tab. Fernando mentioned that too. We can use a mpsc for that to decouple this